### ISSUE TITLE
Compiler complains object is possibly undefined after assigning a value

### ISSUE THREAD

**Issue by yhvicey** [microsoft/TypeScript#31060]
**TypeScript Version:**  3.5.0-dev.20190420

**Search Terms:** `Object is possibly 'undefined'`

**Code**

```ts
interface SomeObject {
    prop: string
}

interface SomeComplexObject {
    prop: SomeObject
}

export class SomeClass {
    private prop: String;

    public constructor(someParam?: SomeObject, someOtherParam?: SomeComplexObject) {
        if (someParam === undefined && someOtherParam === undefined) {
            // <1>
            throw new Error("One of the params must be provided.");
        } else if (someOtherParam !== undefined) {
            // <2>
            someParam = someOtherParam.prop;
        }

        // If someParam === undefined, someOtherParam === undefined, it should goes to <1> and throw;
        // If someParam === undefined, it should goes to <2>, then someParam will be assigned with a non-undefined value, then goes to <3>;
        // If someParam !== undefined, it should goes to <3> directly and should not generate any error.
        // <3>
        this.prop = someParam.prop;
        //              ↑ Object is possibly 'undefined'.ts(2532)
    }
}

```

**Expected behavior:**

`someParam` won't be `undefined` and its props can be accessed

**Actual behavior:**

Compiler complains `Object is possibly 'undefined'`

**Playground Link:** [Playground](https://www.typescriptlang.org/play/#src=interface%20SomeObject%20%7B%0D%0A%20%20%20%20prop%3A%20string%0D%0A%7D%0D%0A%0D%0Ainterface%20SomeComplexObject%20%7B%0D%0A%20%20%20%20prop%3A%20SomeObject%0D%0A%7D%0D%0A%0D%0Aexport%20class%20SomeClass%20%7B%0D%0A%20%20%20%20private%20prop%3A%20String%3B%0D%0A%0D%0A%20%20%20%20public%20constructor(someParam%3F%3A%20SomeObject%2C%20someOtherParam%3F%3A%20SomeComplexObject)%20%7B%0D%0A%20%20%20%20%20%20%20%20if%20(someParam%20%3D%3D%3D%20undefined%20%26%26%20someOtherParam%20%3D%3D%3D%20undefined)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20%3C1%3E%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20throw%20new%20Error(%22One%20of%20the%20params%20must%20be%20provided.%22)%3B%0D%0A%20%20%20%20%20%20%20%20%7D%20else%20if%20(someOtherParam%20!%3D%3D%20undefined)%20%7B%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20%2F%2F%20%3C2%3E%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20someParam%20%3D%20someOtherParam.prop%3B%0D%0A%20%20%20%20%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20If%20someParam%20%3D%3D%3D%20undefined%2C%20someOtherParam%20%3D%3D%3D%20undefined%2C%20it%20should%20goes%20to%20%3C1%3E%20and%20throw%3B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20If%20someParam%20%3D%3D%3D%20undefined%2C%20it%20should%20goes%20to%20%3C2%3E%2C%20then%20someParam%20will%20be%20assigned%20with%20a%20non-undefined%20value%2C%20then%20goes%20to%20%3C3%3E%3B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20If%20someParam%20!%3D%3D%20undefined%2C%20it%20should%20goes%20to%20%3C3%3E%20directly%20and%20should%20not%20generate%20any%20error.%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20%3C3%3E%0D%0A%20%20%20%20%20%20%20%20this.prop%20%3D%20someParam.prop%3B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20%20%20%20%20%20%20%20%20%20%20%20%20%20%E2%86%91%20Object%20is%20possibly%20'undefined'.ts(2532)%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A) (Need to check strictNullCheck)

**Related Issues:** Not yet

**Comment by RyanCavanaugh** [comment]
Our flow control analysis isn't built to detect this kind of complementarity.

The equivalent form
```ts
        if (someParam === undefined) {
            if (someOtherParam === undefined) throw new Error("One of the params must be provided.");
            someParam = someOtherParam.prop;
        }
```
doesn't produce an error.

**Comment by yhvicey** [comment]
In fact, below snippet won't produce an error too:

```ts
interface SomeObject {
    prop: string
}

interface SomeComplexObject {
    prop: SomeObject
}

export class SomeClass {
    private prop: String;

    public constructor(someParam?: SomeObject, someOtherParam?: SomeComplexObject) {
        if (someParam !== undefined) {
            // <1>
            // No op
        } else if (someOtherParam !== undefined) {
            // <2>
            someParam = someOtherParam.prop;
        } else {
            throw new Error("One of the params must be provided.");
        }

        // If someParam === undefined, someOtherParam === undefined, it should goes to <1> and throw;
        // If someParam === undefined, it should goes to <2>, then someParam will be assigned with a non-undefined value, then goes to <3>;
        // If someParam !== undefined, it should goes to <3> directly and should not generate any error.
        // <3>
        this.prop = someParam.prop;
        // More statements using someParam...
    }
}
```

Since I have ~5-6 statements using `someParam` and I prefer not to put them in an if-else block, I will use this one to workaround.

**Comment by dcsan** [comment]
![image](https://user-images.githubusercontent.com/514002/63661164-5f171d80-c7b1-11e9-97bd-70b01c0450bf.png)

why does the above check not prevent this error?

**Comment by yhvicey** [comment]
> ![image](https://user-images.githubusercontent.com/514002/63661164-5f171d80-c7b1-11e9-97bd-70b01c0450bf.png)
> 
> why does the above check not prevent this error?
> 
> ah workaround `tx.btc_volume!`

Guess the compiler is complaining about `tx.btc_volume` might be undefined, not `tx`.

**Comment by dcsan** [comment]
@yhvicey yes you were right. If i put the bang in the right place it shuts up `tx.btc_volume!`

**Comment by dcsan** [comment]
I guess this will be a TS difference from the new JS feature with `tx?.btc_volume?` to fail out and not have to keep checking nulls. like coffeescript had many years ago...

**Comment by surabhi-sadhwani** [comment]
In your tsconfig.js 
 "strictNullChecks": false,
    "strict": false,
https://github.com/Microsoft/TypeScript/issues/31060

**Comment by hari1928** [comment]
look i have an angular code like this 
`this.titleService.setTitle(this.SinglePost[0]['title']['rendered']);`

and its sowing me an error like 
###  ERROR in src/app/blog/blog.component.ts(93,40): error TS2532: Object is possibly 'undefined'.
src/app/blog/blog.component.ts(93,56): error TS2339: Property '0' does not exist on type '[]'.
 
so i got a solution of it 
```
var index =0;
      this.titleService.setTitle(this.SinglePost[index]['title']['rendered']);
```
 and now its working fine

**Comment by Tonterias** [comment]
I get an error saying: src/main/webapp/app/entities/local/newmenu-update.component.html(1,18137): Object is possibly 'null'.  (when I deploy to heroku)

But there is no line number 18137 in newmenu-update.component.html, so the question is: How can I find out where the problem is? I get like a 50+ of this errors in the same page.

**Comment by Yegorich555** [comment]
Also the issue exists when we have the following code:
```ts
function someChecking(value?: {}) {
   const rules  = { key1: "v1", key2: "v2"}
   if (!value) {
       return false;
    }

   // allright this is nested function
    function isRuleFailed(ruleKey: string): any {
       // I get 'possible undefined' here - but it's wrong because function fired after checking value
       return value[ruleKey];
    }

    return isRuleFailed('required') || Object.keys(value).find(isRuleFailed)
    
}
```

**Comment by suhaibmujahid** [comment]
I have the same issue as @Yegorich555.

```ts
if (progress === undefined) return null;
// Object is possibly 'undefined'.  TS2532
if (progress.total === 0) return something;
```

**Comment by RyanCavanaugh** [comment]
This is not the thread to post every "I got an 'object is possibly undefined' error".

Note that narrowing doesn't occur in unreachable code; if your code cannot be reached by any execution path, then you may see errors as a result.

Please post self-contained bugs if you feel you've encountered an error you shouldn't have, or use Stack Overflow to help understand why you're seeing a particular error. Thanks!
